piv/key.go

```diff
@@ -667,24 +671,82 @@ func (yk *YubiKey) PrivateKey(slot Slot, public crypto.PublicKey, auth KeyAuth)
 	}
 }
 
-type keyECDSA struct {
+// ECDSAPrivateKey is used to access a ECDSA signing key.
+type ECDSAPrivateKey struct {
 	yk   *YubiKey
 	slot Slot
 	pub  *ecdsa.PublicKey
 	auth KeyAuth
 	pp   PINPolicy
 }
 
-func (k *keyECDSA) Public() crypto.PublicKey {
+// Public returns the public key associated with this private key.
+func (k *ECDSAPrivateKey) Public() crypto.PublicKey {
 	return k.pub
 }
 
-func (k *keyECDSA) Sign(rand io.Reader, digest []byte, opts crypto.SignerOpts) ([]byte, error) {
+// Sign implements crypto.Signer.
+func (k *ECDSAPrivateKey) Sign(rand io.Reader, digest []byte, opts crypto.SignerOpts) ([]byte, error) {
 	return k.auth.do(k.yk, k.pp, func(tx *scTx) ([]byte, error) {
 		return ykSignECDSA(tx, k.slot, k.pub, digest)
 	})
 }
 
+// SharedKey performs a Diffie-Hellman key agreement with the peer
+// to produce a shared secret key.
+//
+// Peer's public key must use the same algorithm as the key in
+// this slot, or an error will be returned.
+//
+// Length of the result depends on the types and sizes of the keys
+// used for the operation. Callers should use a cryptographic key
+// derivation function to extract the amount of bytes they need.
+func (k *ECDSAPrivateKey) SharedKey(rand io.Reader, peer crypto.PublicKey, opts crypto.SignerOpts) ([]byte, error) {
+	pub, ok := peer.(*ecdsa.PublicKey)
+	if !ok {
+		return nil, errMismatchingAlgorithms
+	}
+	if pub.Curve.Params().BitSize != k.pub.Curve.Params().BitSize {
+		return nil, errMismatchingAlgorithms
+	}
+	msg := elliptic.Marshal(pub.Curve, pub.X, pub.Y)
+	return k.auth.do(k.yk, k.pp, func(tx *scTx) ([]byte, error) {
+		var alg byte
+		size := k.pub.Params().BitSize
+		switch size {
+		case 256:
+			alg = algECCP256
+		case 384:
+			alg = algECCP384
+		default:
+			return nil, fmt.Errorf("unsupported curve: %d", size)
+		}
+
+		// https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-73-4.pdf#page=118
+		cmd := apdu{
```

Comment:
nit: can you also add a comment linking to

https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-73-4.pdf#page=93

That's what specifies 0x85.